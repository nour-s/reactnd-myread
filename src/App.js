import React from "react";
import "./App.css";
import BookCase from "./BookCase";
import SearchBooks from "./SearchBooks";
import * as BooksAPI from "./BooksAPI";
import { BrowserRouter, Route } from "react-router-dom";
import { debounced } from "./Util";

//Map the shelf id to a friendly name.
const shelfNames = {
	currentlyReading: "Currently Reading",
	wantToRead: "Want To Read",
	read: "Read"
};

class App extends React.Component {
	state = {
		shelfs: [], // the list of shelfs with books inside them
		searchResult: [] // the search results if the Search component runs one
	};

	componentDidMount() {
		// When we first mount the component, let's bring the user books (arranged in shelfs).
		BooksAPI.getAll().then(books =>
			this.setState({ shelfs: this.arrangeInShelfs(books) })
		);
	}

	arrangeInShelfs(books) {
		//define a search shelf if we are searching
		var searchShelfId = "searchShelf";
		//Get all the books that the user has put in shelfs already
		var allBooks = this.state.shelfs.reduce((books, shelf) => {
			books.push(...shelf.books);
			return books;
		}, []);

		//put the books in a shelf according to their shelf property value
		return books.reduce((shelfs, book) => {
			// Find the shelf in the aggregated shelfs array if the book has a shelf id, otherwise find the search shelf.
			var shelf = book.shelf
				? shelfs.find(i => i.id === book.shelf)
				: shelfs.find(i => i.id === searchShelfId);
			//If we haven't found the shelf in the aggregated array:
			if (!shelf) {
				//Create a new shelf with the shelf id specified by the first book found that belongs to that shelf.
				shelf = {
					id: book.shelf || searchShelfId,
					name: shelfNames[book.shelf],
					books: []
				};
				//Add it to the aggregated list of shelfs.
				shelfs.push(shelf);
			}
			// Put the book in the shelf that it belongs to.
			shelf.books.push(book);

			//Now, if the book has no shelf specified (the book came from the search API), let's check if the user already put in a shelf.
			if (!book.shelf)
				//This will help the Book component to select an option from the 'Move to' context menu.
				book.shelf = (
					allBooks.find(b => b.id === book.id) || { shelf: undefined }
				).shelf;

			return shelfs;
		}, []);
	}

	onSearch = debounced(query => {
		if (query === "") {
			this.setState({ searchResult: [] });
			return;
		}

		BooksAPI.search(query).then(books => {
			var shelfs = books.error ? [] : this.arrangeInShelfs(books);
			this.setState({ searchResult: shelfs });
		});
	}, 300);

	// Handle clearing the search results
	onClearSearch = () =>
		this.setState({
			searchResult: []
		});

	// Handle moving the book to the passed shelf.
	onMoveBook = (book, shelfId) => {
		return BooksAPI.update(book, shelfId).then(res => {
			// Reflect the change on the book object passed.
			book.shelf = shelfId;
			BooksAPI.getAll().then(books => {
				this.setState({ shelfs: this.arrangeInShelfs(books) });
			});
		});
	};

	render() {
		return (
			<BrowserRouter>
				<div className="app">
					<Route
						path="/"
						exact
						render={() => (
							<BookCase
								onMoveBook={this.onMoveBook}
								shelfs={this.state.shelfs}
								showSearchLink={true}
							/>
						)}
					/>
					<Route
						path="/search:query?"
						exact
						render={props => (
							<SearchBooks
								shelfs={this.state.searchResult}
								onMoveBook={this.onMoveBook}
								onSearch={this.onSearch}
								onClearSearch={this.onClearSearch}
								initQuery={new URLSearchParams(
									props.location.search
								).get("query")}
							/>
						)}
					/>
				</div>
			</BrowserRouter>
		);
	}
}

export default App;
