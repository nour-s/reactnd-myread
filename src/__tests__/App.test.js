import React from "react";
import renderer from "react-test-renderer";
import App from "../App";
import * as BooksAPI from "../BooksAPI";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router";

describe("App test", () => {
	let onMoveBook;
	let onSearch;
	let onClearSearch;
	let rawComponent;
	let booksget;

	let shelfs = [
		{
			id: "read",
			name: "Read",
			books: [{ id: 1, title: "Title 1", shelf: "read" }]
		},
		{
			id: "wantToRead",
			name: "Want To Read",
			books: [{ id: 2, title: "Title 2", shelf: "wantToRead" }]
		}
	];
	//jest.setTimeout(30000);
	let books = [
		{ id: 1, title: "Title 1", shelf: "read" },
		{ id: 2, title: "Title 2", shelf: "wantToRead" }
	];

	beforeEach(() => {
		// spyOn(BooksAPI, "getAll").and.returnValue(Promise.resolve(books));
		//booksget = spyOn(BooksAPI, "getAll");
		BooksAPI["getAll"] = () => Promise.resolve(books);

		onMoveBook = jest.fn();
		onSearch = jest.fn();
		onClearSearch = jest.fn();

		rawComponent = (
			<MemoryRouter initialEntries={["/", "/search"]} initialIndex={1}>
				<App />
			</MemoryRouter>
		);
	});

	it("App should match snapshot", () => {
		const component = renderer.create(rawComponent);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it("App should have a BookCase component with proper props", () => {
		const wrapper = React.createElement(() => rawComponent);
		let component = mount(wrapper);
		let app = component.find("App");
		let bookCase = component.find("BookCase");

		// app.setState({ shelfs: shelfs });

		expect(bookCase).toBeDefined();
		// expect(bookCase.prop("shelfs")).toEqual(shelfs);
	});
});
