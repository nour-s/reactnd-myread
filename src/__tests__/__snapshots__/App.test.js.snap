// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App test App should match snapshot 1`] = `
<div
  className="app"
>
  <div
    className="list-books"
  >
    <div
      className="list-books-title"
    >
      <h1>
        MyReads
      </h1>
    </div>
    <div
      className="list-books-content"
    >
      <div />
    </div>
    <div
      className="open-search"
    >
      <a
        href="/search"
        onClick={[Function]}
      >
        Add a book
      </a>
    </div>
  </div>
</div>
`;
